src/transforms/kubernetes/mod.rs

```diff
@@ -1 +1,552 @@
+        env
+    ))]
+    MissingNodeName { env: &'static str },
+    #[snafu(display("Unknown metadata fields: {:?}", fields))]
+    UnknownFields { fields: Vec<String> },
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
+/// and writes them to the metadata map.
+struct MetadataClient {
+    fields: Vec<Box<dyn Fn(&Pod) -> Vec<(Atom, Value)> + Send + Sync + 'static>>,
+    metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
+    /// (key of data to be deleted, can be deleted after this point in time)
+    delete_queue: VecDeque<(Bytes, Instant)>,
+    client: WatchClient,
+}
+
+impl MetadataClient {
+    fn new(
+        trans_config: &KubePodMetadata,
+        metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
+        client: WatchClient,
+    ) -> Result<Self, BuildError> {
+        // Select Pod metadata fields that need to be extracted from
+        // Pod and then added to Events.
+        let mut add_fields = trans_config.fields.clone();
+        add_fields.sort();
+        add_fields.dedup();
+
+        let mut fields = Vec::new();
+        let mut unknown = Vec::new();
+        let mut all_fields = all_fields();
+
+        for field in add_fields {
+            if let Some(fun) = all_fields.remove(field.as_str()) {
+                fields.push(fun);
+            } else {
+                unknown.push(field.to_owned());
+            }
+        }
+
+        if unknown.is_empty() {
+            Ok(Self {
+                fields,
+                metadata,
+                client,
+                delete_queue: VecDeque::new(),
+            })
+        } else {
+            Err(BuildError::UnknownFields { fields: unknown })
+        }
+    }
+
+    /// Listens for pod metadata changes and updates metadata map.
+    async fn run(mut self) -> BuildError {
+        let mut version = None;
+        let mut error = None;
+        loop {
+            // Build watcher stream
+            let mut watcher = match self.client.watch_metadata(version.clone(), error.take()) {
+                Ok(watcher) => watcher,
+                Err(error) => return BuildError::WatchStreamBuild { source: error },
+            };
+            info!("Watching Pod metadata.");
+
+            // Watch loop
+            let runtime_error = loop {
+                break match watcher.into_future().compat().await {
+                    Ok((Some(event), tail)) => {
+                        watcher = tail;
+
+                        self.delete_update();
+                        version = self.update(event).or(version);
+
+                        self.metadata.refresh();
+                        continue;
+                    }
+                    Ok((None, _)) => RuntimeError::WatchUnexpectedlyEnded,
+                    Err((err, _)) => err,
+                };
+            };
+
+            warn!(
+                message = "Temporary stoped watching Pod metadata.",
```

Comment:
Original message better describes what is happening because we aren't fetching it but watching it. On all, but one error, we won't get any metadata that we already have, once we resume watching.

Reply:
I think you can assume that the metadata watch stopped because it failed, right?

Reply:
Not necessarily, there are other reasons for reseting the connection besides a failure. For ex. server can say that we need to reset connection, or reset version. And I expect most of the times we will reset the connection for those reasons as they are normal to happen.

Reply:
Ok then the message should say "Temporarily unable to watch pod metadata."